sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,412 @@
+            foreach (HttpHeader header in headers)
+            {
+                if (header.Name.StartsWith(Constants.HeaderNames.MetadataPrefix, StringComparison.InvariantCulture))
+                {
+                    metadata[header.Name.Substring(Constants.HeaderNames.MetadataPrefix.Length)] = header.Value;
+                }
+            }
+
+            return metadata;
+        }
+
+        /// <summary>
+        /// Gets and validates a blob's encryption-related metadata
+        /// </summary>
+        /// <param name="metadata">The blob's metadata</param>
+        /// <returns>The relevant metadata.</returns>
+        private static EncryptionData? GetAndValidateEncryptionData(Metadata metadata)
+        {
+            _ = metadata ?? throw new InvalidOperationException();
+
+            EncryptionData encryptionData;
+            if (!metadata.TryGetValue(EncryptionConstants.ENCRYPTION_DATA_KEY, out string encryptedDataString))
+            {
+                return default;
+            }
+
+            using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(encryptedDataString)))
+            {
+                var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
+                encryptionData = (EncryptionData)serializer.ReadObject(stream);
+            }
+
+            _ = encryptionData.ContentEncryptionIV ?? throw new NullReferenceException();
+            _ = encryptionData.WrappedContentKey.EncryptedKey ?? throw new NullReferenceException();
+
+            // Throw if the encryption protocol on the message doesn't match the version that this client library
+            // understands and is able to decrypt.
+            if (EncryptionConstants.ENCRYPTION_PROTOCOL_V1 != encryptionData.EncryptionAgent.Protocol) {
+                throw EncryptionErrors.BadEncryptionAgent(encryptionData.EncryptionAgent.Protocol);
+            }
+
+            return encryptionData;
+        }
+
+        #region GetContentEncryptionKey
+
+        /// <summary>
+        /// Returns the key encryption key for blob. First tries to get key encryption key from KeyResolver, then
+        /// falls back to IKey stored on this EncryptionPolicy.
+        /// </summary>
+        /// <param name="encryptionData">The encryption data.</param>
+        /// <returns>Encryption key as a byte array.</returns>
+        private Memory<byte> GetContentEncryptionKey(EncryptionData encryptionData)
+        {
+            IKeyEncryptionKey key;
+
+            // If we already have a local key and it is the correct one, use that.
+            if (encryptionData.WrappedContentKey.KeyId == this.LocalKey?.KeyId)
+            {
+                key = this.LocalKey;
+            }
+            // Otherwise, use the resolver.
+            else
+            {
+                key = this.KeyResolver?.Resolve(encryptionData.WrappedContentKey.KeyId)
+                    ?? throw EncryptionErrors.KeyNotFound(encryptionData.WrappedContentKey.KeyId);
```

Comment:
Nit - the error kind of gets lost in here.  Would it be more obvious to add a separate `if (key == null)` check below?